### PR TITLE
Override BufferedInputStream to not sychronize single byte reads in Compressor

### PR DESCRIPTION
**Pull request by original-brownbear**
With biased locking gone, we see some slowness in profiling when we use this stream for single byte reads.
This is a recent regression that is a result of
https://openjdk.org/jeps/374.
-> the sychronization overhead for bulk reads hardly matters, but since we do quite a few sinlge byte reads lets fix this. Some rough benchmark numbers on impact can be found at https://gist.github.com/shipilev/71fd881f7cbf3f87028bf87385e84889

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/compress/Compressor.java [hunk 1]
```diff
@@ -26,7 +26,15 @@ public interface Compressor {
      */
     default StreamInput threadLocalStreamInput(InputStream in) throws IOException {
         // wrap stream in buffer since InputStreamStreamInput doesn't do any buffering itself but does a lot of small reads
-        return new InputStreamStreamInput(new BufferedInputStream(threadLocalInputStream(in), DeflateCompressor.BUFFER_SIZE));
+        return new InputStreamStreamInput(new BufferedInputStream(threadLocalInputStream(in), DeflateCompressor.BUFFER_SIZE) {
+            @Override
+            public int read() throws IOException {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/common/compress/Compressor.java:31]: This definitely needs a comment indicating why it's being overridden

**Comment by original-brownbear** [comment]
Thanks Simon, added comment :)

> Do you know what impact this has on ES itself?

Relatively limited impact I think. This is the only place that I could find where we would read from a stream backed `StreamInput` regularly and performance may matter. There are a couple of other spots where we created a `BufferedInputStream` but none of them look performance critical in any way.

**Comment by original-brownbear** [comment]
@thecoop ping :) this should be a quick one, thanks!

**Comment by original-brownbear** [comment]
@elasticsearchmachine update branch

**Review by thecoop — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Simon!

**Pull request merged**